PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx

```diff
@@ -0,0 +1,381 @@
+		end;
+		
+		[Enum.HumanoidStateType.Running] = function()	
+			local sound = Sounds[SFX.Running]
+			verifyAndSetLoopedForSound(sound,true)
+			stopActiveLoopedSoundsExcept(sound)
+			
+			if Util.HorizontalVelocityMagnitude(Head) > 0.5 then
+				if not sound.IsPlaying then
+					Util.Resume(sound)
+				end
+				setSoundInActiveLooped(sound)
+			else
+				stopActiveLoopedSoundsExcept()
+			end
+		end;
+		
+		[Enum.HumanoidStateType.Swimming] = function()
+			if activeState ~= Enum.HumanoidStateType.Swimming and Util.VerticalVelocityMagnitude(Head) > 0.1 then
+				local splashSound = Sounds[SFX.Splash]
+				splashSound.Volume = Util.Clamp(
+					Util.YForLineGivenXAndTwoPts(
+						Util.VerticalVelocityMagnitude(Head), 
+						100, 0.28, 
+						350, 1),
+					0,1)
+				Util.Play(splashSound)
+			end
+			
+			do
+				local sound = Sounds[SFX.Swimming]
+				verifyAndSetLoopedForSound(sound,true)
+				stopActiveLoopedSoundsExcept(sound)
+				if not sound.IsPlaying then
+					Util.Resume(sound)
+				end
+				setSoundInActiveLooped(sound)
+			end
+		end;
+		
+		[Enum.HumanoidStateType.Climbing] = function()
+			local sound = Sounds[SFX.Climbing]
+			verifyAndSetLoopedForSound(sound,true)
+			if Util.VerticalVelocityMagnitude(Head) > 0.1 then
+				if not sound.IsPlaying then
+					Util.Resume(sound)
+				end				
+				stopActiveLoopedSoundsExcept(sound)
+			else
+				stopActiveLoopedSoundsExcept()
+			end		
+			setSoundInActiveLooped(sound)
+		end;
+		
+		[Enum.HumanoidStateType.Jumping] = function()
+			if activeState == Enum.HumanoidStateType.Jumping then
+				return
+			end		
+			stopActiveLoopedSoundsExcept()
+			local sound = Sounds[SFX.Jumping]
+			verifyAndSetLoopedForSound(sound,false)
+			Util.Play(sound)
+		end;
+		
+		[Enum.HumanoidStateType.GettingUp] = function()
+			stopActiveLoopedSoundsExcept()
+			local sound = Sounds[SFX.GettingUp]
+			verifyAndSetLoopedForSound(sound,false)
+			Util.Play(sound)
+		end;
+		
+		[Enum.HumanoidStateType.Freefall] = function()
+			if activeState == Enum.HumanoidStateType.Freefall then
+				return
+			end
+			local sound = Sounds[SFX.FreeFalling]
+			if sound.Volume ~= 0 then
+				sound.Volume = 0
+			end
+			stopActiveLoopedSoundsExcept()
+		end;
+		
+		[Enum.HumanoidStateType.FallingDown] = function()
+			stopActiveLoopedSoundsExcept()
+		end;
+		
+		[Enum.HumanoidStateType.Landed] = function()
+			stopActiveLoopedSoundsExcept()
+			if Util.VerticalVelocityMagnitude(Head) > 75 then
+				local landingSound = Sounds[SFX.Landing]
+				landingSound.Volume = Util.Clamp(
+					Util.YForLineGivenXAndTwoPts(
+						Util.VerticalVelocityMagnitude(Head), 
+						50, 0, 
+						100, 1),
+					0,1)
+				Util.Play(landingSound)			
+			end
+		end
+	}
+	
+	-- Handle state event fired or OnChange fired
+	function stateUpdated(state)
+		if stateUpdateHandler[state] ~= nil then
+			stateUpdateHandler[state]()
+		end
+		activeState = state
+	end
+	
+	-- Runs on heartbeat
+	function onHeartbeat(step)
```

Comment:
Heartbeat's here for some new behaviour:
1. Only trigger the freefalling sound if the last active state was "FreeFall", and vertical speed is above 75 (arbitrary number that made it NOT trigger for normal jumps). 

2. If last active state was running, and horizontal speed is less than 0.5, stop the running sound playback.

I personally think it's a lot cleaner to do this with a stateless update method (or else we'd need to manage/preempt the spawn or delay functions). The only "state" that this code has right now is "lastActiveState" and "activeLoopedSounds" (soon to be renamed playingLoopedSounds).

And they don't seem to be a big problem performance-wise. Heartbeat stuff generally shouldn't be, is that correct? (It's one heartbeat listener per client too)